File: app/src/main/kotlin/net/tigr/soulcalc/ui/screen/MainScreen.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Albert Zenkoff
 */

package net.tigr.soulcalc.ui.screen

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.widget.Toast
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import net.tigr.soulcalc.R
import net.tigr.soulcalc.data.preferences.ThemeMode
import net.tigr.soulcalc.ui.component.AboutDialog
import net.tigr.soulcalc.ui.component.CalculatorKeyboard
import net.tigr.soulcalc.ui.component.KeyboardToggle
import net.tigr.soulcalc.ui.component.LineRow
import net.tigr.soulcalc.ui.component.SettingsDialog
import net.tigr.soulcalc.ui.component.VariablePicker
import net.tigr.soulcalc.ui.viewmodel.MainUiEvent
import net.tigr.soulcalc.ui.viewmodel.MainViewModel

/**
 * Main screen of the calculator application.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MainScreen(
    viewModel: MainViewModel,
    currentThemeMode: ThemeMode,
    onThemeModeChanged: (ThemeMode) -> Unit,
    modifier: Modifier = Modifier
) {
    val uiState by viewModel.uiState.collectAsState()
    val context = LocalContext.current
    val listState = rememberLazyListState()
    var showMenu by remember { mutableStateOf(false) }
    var showSettings by remember { mutableStateOf(false) }
    var showAbout by remember { mutableStateOf(false) }


    LaunchedEffect(uiState.toastMessage) {
        uiState.toastMessage?.let { message ->
            Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
            viewModel.onEvent(MainUiEvent.ToastShown)
        }
    }

    LaunchedEffect(uiState.focusedLineIndex, uiState.lines.size) {
        if (uiState.focusedLineIndex >= 0 && uiState.focusedLineIndex < uiState.lines.size) {
            listState.animateScrollToItem(uiState.focusedLineIndex)
        }
    }

    if (uiState.showVariablePicker) {
        VariablePicker(
            variables = uiState.availableVariables,
            onVariableSelected = { variable ->
                viewModel.onEvent(MainUiEvent.VariableSelected(variable))
            },
            onDismiss = {
                viewModel.onEvent(MainUiEvent.DismissVariablePicker)
            }
        )
    }

    if (showSettings) {
        SettingsDialog(
            currentThemeMode = currentThemeMode,
            onThemeModeChanged = onThemeModeChanged,
            onDismiss = { showSettings = false }
        )
    }

    if (showAbout) {
        AboutDialog(
            onDismiss = { showAbout = false }
        )
    }

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(stringResource(R.string.app_name)) },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.surface
                ),
                actions = {
                    KeyboardToggle(
                        useCustomKeyboard = uiState.useCustomKeyboard,
                        onToggle = { viewModel.onEvent(MainUiEvent.ToggleKeyboard) }
                    )

                    IconButton(onClick = { showMenu = true }) {
                        Icon(
                            imageVector = Icons.Default.MoreVert,
                            contentDescription = stringResource(R.string.content_description_menu)
                        )
                    }

                    DropdownMenu(
                        expanded = showMenu,
                        onDismissRequest = { showMenu = false }
                    ) {
                        DropdownMenuItem(
                            text = { Text(stringResource(R.string.menu_copy_all)) },
                            onClick = {
                                showMenu = false
                                val text = viewModel.formatAllLines()
                                if (text.isNotEmpty()) {
                                    copyToClipboard(context, text)
                                    viewModel.onEvent(MainUiEvent.CopyAll)
                                }
                            },
                            leadingIcon = {
                                Icon(
                                    imageVector = Icons.Default.ContentCopy,
                                    contentDescription = null
                                )
                            }
                        )
                        DropdownMenuItem(
                            text = { Text(stringResource(R.string.menu_clear_sheet)) },
                            onClick = {
                                showMenu = false
                                viewModel.onEvent(MainUiEvent.ClearSheet)
                            },
                            leadingIcon = {
                                Icon(
                                    imageVector = Icons.Default.Delete,
                                    contentDescription = null
                                )
                            }
                        )
                        HorizontalDivider()
                        DropdownMenuItem(
                            text = { Text(stringResource(R.string.menu_settings)) },
                            onClick = {
                                showMenu = false
                                showSettings = true
                            },
                            leadingIcon = {
                                Icon(
                                    imageVector = Icons.Default.Settings,
                                    contentDescription = null
                                )
                            }
                        )
                        DropdownMenuItem(
                            text = { Text(stringResource(R.string.menu_about)) },
                            onClick = {
                                showMenu = false
                                showAbout = true
                            },
                            leadingIcon = {
                                Icon(
                                    imageVector = Icons.Default.Info,
                                    contentDescription = null
                                )
                            }
                        )
                    }
                }
            )
        }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .imePadding()
        ) {
            LazyColumn(
                state = listState,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxWidth()
            ) {
                itemsIndexed(
                    items = uiState.lines,
                    key = { index, line -> "${line.id}_$index" }
                ) { index, line ->
                    LineRow(
                        line = line,
                        lineNumber = index + 1,
                        isFocused = index == uiState.focusedLineIndex,
                        pendingInsertion = if (index == uiState.focusedLineIndex) {
                            uiState.pendingInsertion
                        } else null,
                        pendingBackspace = index == uiState.focusedLineIndex && uiState.pendingBackspace,
                        pendingClearLine = index == uiState.focusedLineIndex && uiState.pendingClearLine,
                        showSystemKeyboard = !uiState.useCustomKeyboard,
                        onTextChanged = { text ->
                            viewModel.onEvent(MainUiEvent.LineTextChanged(index, text))
                        },
                        onNewLine = {
                            viewModel.onEvent(MainUiEvent.NewLineRequested(index))
                        },
                        onDeleteLine = {
                            viewModel.onEvent(MainUiEvent.DeleteLineRequested(index))
                        },
                        onResultTap = {
                            viewModel.getResultForClipboard(index)?.let { result ->
                                copyToClipboard(context, result)
                                viewModel.onEvent(MainUiEvent.CopyResult(index))
                            }
                        },
                        onFocused = {
                            viewModel.onEvent(MainUiEvent.LineFocused(index))
                        },
                        onInsertionConsumed = {
                            viewModel.onEvent(MainUiEvent.InsertionConsumed)
                        },
                        onBackspaceConsumed = {
                            viewModel.onEvent(MainUiEvent.BackspaceConsumed)
                        },
                        onClearLineConsumed = {
                            viewModel.onEvent(MainUiEvent.ClearLineConsumed)
                        }
                    )
                }
            }

            if (uiState.useCustomKeyboard) {
                CalculatorKeyboard(
                    onKeyPress = { key ->
                        viewModel.onEvent(MainUiEvent.KeyPressed(key))
                    },
                    onEnter = {
                        viewModel.onEvent(MainUiEvent.EnterKeyPressed(uiState.focusedLineIndex))
                    },
                    onBackspace = {
                        viewModel.onEvent(MainUiEvent.BackspacePressed)
                    },
                    onClearLine = {
                        viewModel.onEvent(MainUiEvent.ClearLinePressed)
                    },
                    onDollarKey = {
                        viewModel.onEvent(MainUiEvent.DollarKeyPressed)
                    },
                    onDollarLongPress = {
                        viewModel.onEvent(MainUiEvent.ShowVariablePicker)
                    },
                    onHashKey = {
                        viewModel.onEvent(MainUiEvent.HashKeyPressed)
                    },
                    modifier = Modifier.fillMaxWidth()
                )
            }
        }
    }
}

private fun copyToClipboard(context: Context, text: String) {
    val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    val clip = ClipData.newPlainText("SoulCalc", text)
    clipboard.setPrimaryClip(clip)
}

File: ios/SoulCalc/UI/Screen/MainScreen.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Albert Zenkoff

import SwiftUI
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

/// Main screen of the calculator application.
struct MainScreen: View {
    @ObservedObject var viewModel: MainViewModel
    @Binding var themeMode: ThemeMode

    @State private var showMenu = false
    @State private var showSettings = false
    @State private var showGuide = false
    @State private var showAbout = false

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                // Lines list
                ScrollViewReader { proxy in
                    ScrollView {
                        LazyVStack(spacing: 0) {
                            ForEach(Array(viewModel.state.lines.enumerated()), id: \.offset) { index, line in
                                LineRow(
                                    line: line,
                                    lineNumber: index + 1,
                                    isFocused: index == viewModel.state.focusedLineIndex,
                                    pendingInsertion: index == viewModel.state.focusedLineIndex
                                        ? viewModel.state.pendingInsertion : nil,
                                    pendingBackspace: index == viewModel.state.focusedLineIndex
                                        && viewModel.state.pendingBackspace,
                                    pendingClearLine: index == viewModel.state.focusedLineIndex
                                        && viewModel.state.pendingClearLine,
                                    showSystemKeyboard: !viewModel.state.useCustomKeyboard,
                                    onTextChanged: { text in
                                        viewModel.lineTextChanged(index, text)
                                    },
                                    onNewLine: {
                                        viewModel.newLineRequested(index)
                                    },
                                    onDeleteLine: {
                                        viewModel.deleteLineRequested(index)
                                    },
                                    onResultTap: {
                                        if let result = viewModel.copyResult(index) {
                                            #if canImport(UIKit)
                                            UIPasteboard.general.string = result
                                            #elseif canImport(AppKit)
                                            NSPasteboard.general.clearContents()
                                            NSPasteboard.general.setString(result, forType: .string)
                                            #endif
                                        }
                                    },
                                    onFocused: {
                                        viewModel.lineFocused(index)
                                    },
                                    onInsertionConsumed: {
                                        viewModel.insertionConsumed()
                                    },
                                    onBackspaceConsumed: {
                                        viewModel.backspaceConsumed()
                                    },
                                    onClearLineConsumed: {
                                        viewModel.clearLineConsumed()
                                    }
                                )
                                .id(index)
                            }
                        }
                    }
                    .onChange(of: viewModel.state.focusedLineIndex) { _, newIndex in
                        withAnimation {
                            proxy.scrollTo(newIndex, anchor: .center)
                        }
                    }
                    .onChange(of: viewModel.state.lines.count) { _, _ in
                        withAnimation {
                            proxy.scrollTo(viewModel.state.focusedLineIndex, anchor: .center)
                        }
                    }
                }

                // Custom keyboard
                if viewModel.state.useCustomKeyboard {
                    CalculatorKeyboard(
                        onKeyPress: { key in viewModel.keyPressed(key) },
                        onEnter: { viewModel.enterKeyPressed() },
                        onBackspace: { viewModel.backspacePressed() },
                        onClearLine: { viewModel.clearLinePressed() },
                        onDollarKey: { viewModel.dollarKeyPressed() },
                        onDollarLongPress: { viewModel.showVariablePicker() },
                        onHashKey: { viewModel.hashKeyPressed() }
                    )
                }
            }
            .navigationTitle("SoulCalc")
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .primaryAction) {
                    HStack(spacing: 4) {
                        // Keyboard toggle
                        Button {
                            viewModel.toggleKeyboard()
                        } label: {
                            Image(systemName: viewModel.state.useCustomKeyboard
                                  ? "keyboard" : "keyboard.chevron.compact.down")
                        }

                        // Menu
                        Menu {
                            Button {
                                let text = viewModel.copyAll()
                                if !text.isEmpty {
                                    #if canImport(UIKit)
                                    UIPasteboard.general.string = text
                                    #elseif canImport(AppKit)
                                    NSPasteboard.general.clearContents()
                                    NSPasteboard.general.setString(text, forType: .string)
                                    #endif
                                }
                            } label: {
                                Label("Copy All", systemImage: "doc.on.doc")
                            }

                            Button(role: .destructive) {
                                viewModel.clearSheet()
                            } label: {
                                Label("Clear Sheet", systemImage: "trash")
                            }

                            Divider()

                            Button {
                                showGuide = true
                            } label: {
                                Label("Guide", systemImage: "questionmark.circle")
                            }

                            Button {
                                showSettings = true
                            } label: {
                                Label("Settings", systemImage: "gearshape")
                            }

                            Button {
                                showAbout = true
                            } label: {
                                Label("About", systemImage: "info.circle")
                            }
                        } label: {
                            Image(systemName: "ellipsis.circle")
                        }
                    }
                }
            }
            .sheet(isPresented: $showSettings) {
                SettingsView(currentThemeMode: $themeMode)
                    .presentationDetents([.medium])
            }
            .sheet(isPresented: $showGuide) {
                GuideView()
            }
            .sheet(isPresented: $showAbout) {
                AboutView()
                    .presentationDetents([.medium])
            }
            .sheet(isPresented: Binding(
                get: { viewModel.state.showVariablePicker },
                set: { if !$0 { viewModel.dismissVariablePicker() } }
            )) {
                VariablePickerView(
                    variables: viewModel.state.availableVariables,
                    onVariableSelected: { variable in
                        viewModel.variableSelected(variable)
                    },
                    onDismiss: {
                        viewModel.dismissVariablePicker()
                    }
                )
                .presentationDetents([.medium])
            }
        }
        .overlay(alignment: .bottom) {
            // Toast
            if let message = viewModel.state.toastMessage {
                Text(message)
                    .font(.subheadline)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 10)
                    .background(.ultraThinMaterial, in: Capsule())
                    .padding(.bottom, viewModel.state.useCustomKeyboard ? 300 : 60)
                    .transition(.move(edge: .bottom).combined(with: .opacity))
                    .onAppear {
                        DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
                            withAnimation {
                                viewModel.toastShown()
                            }
                        }
                    }
            }
        }
        .animation(.easeInOut(duration: 0.2), value: viewModel.state.toastMessage != nil)
    }
}
